### ISSUE TITLE
O_CLOEXEC undefined in glfw source on Linux release builds

### ISSUE THREAD

**Issue by nleseul** [johnnovak/gridmonger#19]
It looks like this is the only issue I'm having with building on Linux at this point! Workaround is described below. Just documenting my experience in case it helps someone more familiar with the nim build system come up with a real fix.

### Steps to reproduce

Attempt to build on a Linux system with `nim release -d:osdialogGtk3` (or any osdialog define).

(My system is Linux Mint 22... let me know if you need me to dig up the versions of gcc, nim, or any other library/package.)

### Results

Compile errors similar to the following:

```
/home/<user>/.nimble/pkgs2/glfw-3.4.0.4-7a22b020d0fcae4174ad6edcf611339b84924f76/glfw/src/linux_joystick.c: In function ‘openJoystickDevice’:
/home/<user>/.nimble/pkgs2/glfw-3.4.0.4-7a22b020d0fcae4174ad6edcf611339b84924f76/glfw/src/linux_joystick.c:138:51: error: ‘O_CLOEXEC’ undeclared (first use in this function); did you mean ‘IN_CLOEXEC’?
  138 |     linjs.fd = open(path, O_RDONLY | O_NONBLOCK | O_CLOEXEC);
      |                                                   ^~~~~~~~~
      |                                                   IN_CLOEXEC
/home/<user>/.nimble/pkgs2/glfw-3.4.0.4-7a22b020d0fcae4174ad6edcf611339b84924f76/glfw/src/linux_joystick.c:138:51: note: each undeclared identifier is reported only once for each function it appears in
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3 -pthread -D_GLFW_X11  -I/home/<user>/.nimble/pkgs2/nanovg-0.4.0-08ff8efaeb1e09af20b11cab53ef9eb8282fc122/nanovg/deps/stb -DNANOVG_GL3_IMPLEMENTATION -g -Wall -Wextra -std=c99 -pedantic -I/usr/include/gtk-3.0 -I/usr/include/pango-1.0 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/fribidi -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/x86_64-linux-gnu -I/usr/include/webp -I/usr/include/gio-unix-2.0 -I/usr/include/atk-1.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -pthread -O3 -fno-strict-aliasing -fno-ident -fno-math-errno   -I/home/<user>/.choosenim/toolchains/nim-2.0.8/lib -I/home/<user>/Development/External/gridmonger/src -o /home/<user>/.cache/nim/main_r/@m..@s..@s..@s..@s.nimble@spkgs2@sglfw-3.4.0.4-7a22b020d0fcae4174ad6edcf611339b84924f76@sglfw@ssrc@slinux_joystick.c.o /home/<user>/.nimble/pkgs2/glfw-3.4.0.4-7a22b020d0fcae4174ad6edcf611339b84924f76/glfw/src/linux_joystick.c' failed with exit code: 1
```

### Workaround

If I copy and paste the `gcc` command line given in the error, but remove the option `-std=c99`, then that command executes successfully. Repeating the `nim release` command afterwards allows the gridmonger build to finish with no further issues (since the manual compilation puts an object file in the package cache which the linker can grab without needing to try to compile this file again).

### Additional notes

This only occurs when building release; the debug build succeeds without issue. Building the examples in the nim-glfw repo also succeeds without issue. As far as I can tell, the `-std=c99` flag mentioned above doesn't get used in either of those situations; it only gets added to the command line during the gridmonger release build. (I haven't been able to figure out why that is.)

**Comment by Daerandin** [comment]
You can override this while building gridmonger by using the -t flag. For example:

`nim release -d:osdialogGtk3 -t:gnu18`

I have not yet done any extensive testing to see if this causes any errors. gnu18 is the default for gcc for C code, which is why I just tried with this setting.

**Comment by johnnovak** [comment]
Thanks @nleseul @Daerandin. I will provide Linux x64 and ARM binaries... on day 😄 I just find the whole Linux packaging situation overwhelming, so it will be two tar.gz file with the binaries. I'll work it out on my RPi 4 box when I'm bored 😄 

Does the custom drawn windows decoration stuff work for you guys? Can you move around the window and resize it by dragging either of its four corners and sides without getting horrible flicker? One guy reported the flicker a year ago, but he was using some tiling window manager. Maybe I should introduce a "use standard window decorations" option anyway as fallback.

**Comment by Daerandin** [comment]
I can resize the window without any issues, no flicker. I can't move it around with the custom window decoration, but I can move it by using the keyboard shortcuts for my window manager (sway), so the issue seems to be the custom window decorations. I don't know if the behavior would be the same in a normal desktop environment.

Personally I prefer to just compile it myself so I'll probably stick to that even if you provide official Linux builds.

Also I noticed that I made a mistake in the compile line in my previous comment, fixed it with an edit.

**Comment by johnnovak** [comment]
> I can resize the window without any issues, no flicker. I can't move it around with the custom window decoration, but I can move it by using the keyboard shortcuts for my window manager (sway), so the issue seems to be the custom window decorations. I don't know if the behavior would be the same in a normal desktop environment.

Yeah, I had to massage the custom window stuff so it works equally well on Windows and macOS. Probably it wouldn't be too hard to get it to work on KDE, but that's where I draw the line 😆 I think we need that option to revert to standard windows decorations for the 10000 different Linux windows managers out there...

> Personally I prefer to just compile it myself so I'll probably stick to that even if you provide official Linux builds.

Of course, and that's completely fine! Just I'll need to have a single self-documenting Linux workflow action that builds the x84 and ARM binaries.

I already have those for macOS and Windows:
https://github.com/johnnovak/gridmonger/actions/runs/12021506923

Then I'll at least know if I broke the "official Linux build" method by mistake 😄 
 
> Also I noticed that I made a mistake in the compile line in my previous comment, fixed it with an edit.

Thanks man!

**Comment by johnnovak** [comment]
@nleseul Thanks for that info. Yes, the custom window drawing stuff is highly dependent on the exact order of windowing system events, so it must be tweaked per OS and window manager combo. That's doable for Windows and macOS, but a potential nightmare on Linux with all the options.

I might tweak it so it works with the default PIXEL environment of my Raspberry Pi Linux box, then maybe KDE Plasma and Xcfe if I'm feeling adventurous 😆
